R/foreign-keys.R

```diff
@@ -189,20 +192,54 @@ cdm_enum_fk_candidates <- nse_function(c(dm, table, ref_table), ~ {
 
   tbl <- cdm_get_tables(dm)[[table_name]]
   tbl_colnames <- colnames(tbl)
+  # classes of columns from 'table'?
+  cols_classes <- map_chr(
+    tbl_colnames,
+    ~class(head(tbl, 0) %>% pull(!!sym(.x)))[[1]]
+  ) %>% set_names(tbl_colnames)
 
+  # what `class` is pk-column of ref_table?
   ref_tbl <- cdm_get_tables(dm)[[ref_table_name]]
+  class_pk <- class(head(ref_tbl, 0) %>% pull(!!sym(ref_tbl_pk)))[[1]]
+
+  # which classes match/don't match?
+  cols_class_match <- tbl_colnames[cols_classes == class_pk]
+  tibble_cols_class_no_match <- tibble(
+    column = setdiff(tbl_colnames, cols_class_match),
+    candidate = FALSE,
+    why = map_chr(
+      column,
+      ~glue("class `{cols_classes[.x]}` differs from PK-column class `{class_pk}`")
+      )
+    )
 
-  subsets <- map_lgl(
-    tbl_colnames,
-    ~ is_subset(tbl, !!.x, ref_tbl, !!ref_tbl_pk)
-  )
+  # which columns contain subset of values of PK-col?
+  col_candidates <- cols_class_match[map_lgl(
+    cols_class_match,
+    ~ is_subset(tbl, !!.x, ref_tbl, !!ref_tbl_pk))]
+  tibble_candidates <- tibble(column = col_candidates,
+    candidate = TRUE,
+    why = "")
+
+  # which columns of same class contain additional values and which ones?
+  set_of_pk_vals <- pull(arrange(distinct(ref_tbl, !!ensym(ref_tbl_pk))))
```

Comment:
Can you please use the following idiom to compute mismatches:

```r
tryCatch(
  anti_join(...) %>%
     utils::head(MAX_COMMAS + 1) %>%
     collect(),
  error = identity
)
```

Then, check for `is_condition(.)` and, if yes, use `conditionMessage(.)` as reason.